File: 649-dota2-senate/649-dota2-senate.java
class Solution {
    public String predictPartyVictory(String senate) {
        char[] senates = senate.toCharArray();
        int n = senates.length;
        int radiant = 0, dire = 0;
        
        for(int i = 0; i < n;i++) {
            if(senates[i] == 'R') radiant++;
            else dire++;
        }
        int start = 0;
        while(radiant > 0 && dire > 0){
            if(senates[start] != 'R' && senates[start] != 'D'){
                start = (start + 1) % n;
            }
            else{
                char currBanned = 'R'; 
                if(senates[start] == 'R'){
                    currBanned = 'D';
                    dire--;
                }else radiant--;
                int nxt = (start + 1) % n;
                start = nxt;
                while(senates[nxt] != currBanned){
                    nxt = (nxt + 1) % n;
                }
                senates[nxt] = ' ';
            }
        }
        
        return (radiant > 0) ? "Radiant" : "Dire";
    }
}

/*
st: 3:37 
sub: 4:16 
*/

File: 649-dota2-senate/649-dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        
        r_banned = 0 
        d_banned = 0 
        r_count = senate.count('R')
        d_count = senate.count('D')
        
        senate = list(senate)
        while r_count > 0 and d_count > 0: 
            for i,s in enumerate(senate): 
                if s == 'R':
                    if r_banned > 0: 
                        senate[i] = " "
                        r_banned -= 1 
                        r_count -= 1 
                    else:
                        d_banned += 1 
                elif s == 'D':
                    if d_banned > 0:
                        senate[i] = " "
                        d_count -= 1 
                        d_banned -= 1 
                    else: 
                        r_banned += 1
                        
        if r_count: 
            return "Radiant" 
        else:
            return "Dire"
            
                
        
        """
        
        Rights: 
            -> Ban one (Remove elem) 
            -> Announce 
        """